File: basketball-stats-api/src/main/java/com/hoops/stats/card/service/RatingCalculator.java
package com.hoops.stats.card.service;

import com.hoops.stats.model.PlayerStats;
import org.springframework.stereotype.Component;

import java.util.HashMap;
import java.util.Map;

/**
 * Rating Calculator
 * Calculates FIFA Ultimate Team style overall ratings (0-99) for players
 */
@Component
public class RatingCalculator {

    /**
     * Calculate overall rating for a player
     */
    public int calculateOverallRating(PlayerStats stats, String position) {
        Map<String, Double> normalized = normalizeStats(stats);
        Map<String, Double> weights = getPositionWeights(position);

        double totalWeight = 0;
        double weightedSum = 0;

        for (Map.Entry<String, Double> entry : weights.entrySet()) {
            String stat = entry.getKey();
            double weight = entry.getValue();

            if (normalized.containsKey(stat)) {
                weightedSum += normalized.get(stat) * weight;
                totalWeight += weight;
            }
        }

        double rawRating = totalWeight > 0 ? weightedSum / totalWeight : 50;
        double adjustedRating = applyPositionAdjustments(rawRating, stats);

        // Clamp to 40-99 range
        return (int) Math.round(Math.max(40, Math.min(99, adjustedRating)));
    }

    /**
     * Normalize stats to 0-100 scale
     */
    private Map<String, Double> normalizeStats(PlayerStats stats) {
        Map<String, Double> normalized = new HashMap<>();

        // Scoring (0-40 PPG range)
        normalized.put("points", normalizeValue(stats.getPointsPerGame(), 0, 35, 0, 100));

        // Playmaking (0-15 APG range)
        normalized.put("assists", normalizeValue(stats.getAssistsPerGame(), 0, 12, 0, 100));

        // Rebounding (0-15 RPG range)
        normalized.put("rebounds", normalizeValue(stats.getReboundsPerGame(), 0, 15, 0, 100));

        // Defense
        normalized.put("steals", normalizeValue(stats.getStealsPerGame(), 0, 3, 0, 100));
        normalized.put("blocks", normalizeValue(stats.getBlocksPerGame(), 0, 3, 0, 100));

        // Efficiency
        normalized.put("fieldGoalPct", normalizeValue(stats.getFieldGoalPercentage(), 0, 70, 0, 100));
        normalized.put("threePointPct", normalizeValue(stats.getThreePointPercentage(), 0, 50, 0, 100));
        normalized.put("freeThrowPct", normalizeValue(stats.getFreeThrowPercentage(), 0, 95, 0, 100));

        return normalized;
    }

    /**
     * Get position-specific stat weights
     */
    private Map<String, Double> getPositionWeights(String position) {
        Map<String, Double> weights = new HashMap<>();

        switch (position != null ? position.toUpperCase() : "SF") {
            case "PG": // Point Guard - assists and ball handling
                weights.put("points", 0.20);
                weights.put("assists", 0.30);
                weights.put("rebounds", 0.10);
                weights.put("steals", 0.15);
                weights.put("blocks", 0.05);
                weights.put("fieldGoalPct", 0.10);
                weights.put("threePointPct", 0.05);
                weights.put("freeThrowPct", 0.05);
                break;

            case "SG": // Shooting Guard - scoring and perimeter shooting
                weights.put("points", 0.30);
                weights.put("assists", 0.15);
                weights.put("rebounds", 0.10);
                weights.put("steals", 0.15);
                weights.put("blocks", 0.05);
                weights.put("fieldGoalPct", 0.10);
                weights.put("threePointPct", 0.10);
                weights.put("freeThrowPct", 0.05);
                break;

            case "SF": // Small Forward - balanced all-around
                weights.put("points", 0.25);
                weights.put("assists", 0.15);
                weights.put("rebounds", 0.15);
                weights.put("steals", 0.15);
                weights.put("blocks", 0.10);
                weights.put("fieldGoalPct", 0.10);
                weights.put("threePointPct", 0.05);
                weights.put("freeThrowPct", 0.05);
                break;

            case "PF": // Power Forward - rebounds and interior scoring
                weights.put("points", 0.25);
                weights.put("assists", 0.10);
                weights.put("rebounds", 0.25);
                weights.put("steals", 0.05);
                weights.put("blocks", 0.15);
                weights.put("fieldGoalPct", 0.15);
                weights.put("threePointPct", 0.00);
                weights.put("freeThrowPct", 0.05);
                break;

            case "C": // Center - rebounds and blocks
                weights.put("points", 0.20);
                weights.put("assists", 0.05);
                weights.put("rebounds", 0.30);
                weights.put("steals", 0.05);
                weights.put("blocks", 0.20);
                weights.put("fieldGoalPct", 0.15);
                weights.put("threePointPct", 0.00);
                weights.put("freeThrowPct", 0.05);
                break;

            default:
                // Default to SF weights
                return getPositionWeights("SF");
        }

        return weights;
    }

    /**
     * Apply position-specific adjustments
     */
    private double applyPositionAdjustments(double rating, PlayerStats stats) {
        double adjusted = rating;

        // Bonus for elite efficiency
        if (stats.getFieldGoalPercentage() != null && stats.getFieldGoalPercentage() > 55) {
            adjusted += 2;
        }

        // Bonus for high volume scorers
        if (stats.getPointsPerGame() != null) {
            if (stats.getPointsPerGame() > 28) {
                adjusted += 3;
            } else if (stats.getPointsPerGame() > 25) {
                adjusted += 2;
            }
        }

        // Bonus for triple-double threats
        if (stats.getPointsPerGame() != null && stats.getAssistsPerGame() != null &&
                stats.getReboundsPerGame() != null) {
            if (stats.getPointsPerGame() > 15 && stats.getAssistsPerGame() > 7 &&
                    stats.getReboundsPerGame() > 7) {
                adjusted += 4;
            }
        }

        // Penalty for low efficiency
        if (stats.getFieldGoalPercentage() != null && stats.getPointsPerGame() != null) {
            if (stats.getFieldGoalPercentage() < 40 && stats.getPointsPerGame() > 15) {
                adjusted -= 3;
            }
        }

        return adjusted;
    }

    /**
     * Determine card rarity based on overall rating
     */
    public String getCardRarity(int overallRating) {
        if (overallRating >= 90) return "legendary";
        if (overallRating >= 85) return "gold_rare";
        if (overallRating >= 80) return "gold";
        if (overallRating >= 75) return "silver_rare";
        if (overallRating >= 65) return "silver";
        return "bronze";
    }

    /**
     * Get card color based on rarity
     */
    public String getCardColor(String rarity) {
        return switch (rarity) {
            case "bronze" -> "#CD7F32";
            case "silver" -> "#C0C0C0";
            case "silver_rare" -> "#E8E8E8";
            case "gold" -> "#FFD700";
            case "gold_rare" -> "#FFA500";
            case "legendary" -> "#9C27B0";
            default -> "#CD7F32";
        };
    }

    /**
     * Get gradient colors for card background
     */
    public String[] getCardGradient(String rarity) {
        return switch (rarity) {
            case "bronze" -> new String[]{"#8B4513", "#CD7F32", "#A0522D"};
            case "silver" -> new String[]{"#808080", "#C0C0C0", "#A9A9A9"};
            case "silver_rare" -> new String[]{"#C0C0C0", "#E8E8E8", "#D3D3D3"};
            case "gold" -> new String[]{"#DAA520", "#FFD700", "#FFA500"};
            case "gold_rare" -> new String[]{"#FF8C00", "#FFD700", "#FF6347"};
            case "legendary" -> new String[]{"#4A148C", "#9C27B0", "#E91E63"};
            default -> new String[]{"#8B4513", "#CD7F32", "#A0522D"};
        };
    }

    /**
     * Normalize value to target range
     */
    private double normalizeValue(Double value, double sourceMin, double sourceMax,
                                   double targetMin, double targetMax) {
        if (value == null) return targetMin;

        // Clamp source value
        double clamped = Math.max(sourceMin, Math.min(sourceMax, value));

        // Normalize to 0-1
        double normalized = (clamped - sourceMin) / (sourceMax - sourceMin);

        // Scale to target range
        return targetMin + normalized * (targetMax - targetMin);
    }
}

File: scripts/ratingCalculator.js
/**
 * Overall Rating Calculator
 * Converts basketball stats into FIFA Ultimate Team style ratings (0-99)
 */

/**
 * Calculate overall rating for a player based on position and stats
 * @param {Object} stats - Player statistics
 * @param {string} position - Player position (PG, SG, SF, PF, C)
 * @returns {number} Overall rating (0-99)
 */
function calculateOverallRating(stats, position) {
    // Normalize stats to 0-100 scale
    const normalized = normalizeStats(stats);

    // Position-specific weights
    const weights = getPositionWeights(position);

    // Calculate weighted average
    let totalWeight = 0;
    let weightedSum = 0;

    for (const [stat, weight] of Object.entries(weights)) {
        if (normalized[stat] !== undefined) {
            weightedSum += normalized[stat] * weight;
            totalWeight += weight;
        }
    }

    const rawRating = totalWeight > 0 ? weightedSum / totalWeight : 50;

    // Apply position bonus/penalty
    const adjustedRating = applyPositionAdjustments(rawRating, stats, position);

    // Clamp to 40-99 range (no player below 40, max is 99)
    return Math.round(Math.max(40, Math.min(99, adjustedRating)));
}

/**
 * Normalize stats to 0-100 scale based on NBA benchmarks
 */
function normalizeStats(stats) {
    return {
        // Scoring (0-40 PPG range)
        points: normalizeValue(stats.ppg || stats.pointsPerGame || 0, 0, 35, 0, 100),

        // Playmaking (0-15 APG range)
        assists: normalizeValue(stats.apg || stats.assistsPerGame || 0, 0, 12, 0, 100),

        // Rebounding (0-15 RPG range)
        rebounds: normalizeValue(stats.rpg || stats.reboundsPerGame || 0, 0, 15, 0, 100),

        // Defense
        steals: normalizeValue(stats.spg || stats.stealsPerGame || 0, 0, 3, 0, 100),
        blocks: normalizeValue(stats.bpg || stats.blocksPerGame || 0, 0, 3, 0, 100),

        // Efficiency
        fieldGoalPct: normalizeValue(parsePercentage(stats.fgPct || stats.fieldGoalPercentage || 0), 0, 70, 0, 100),
        threePointPct: normalizeValue(parsePercentage(stats.fg3Pct || stats.threePointPercentage || 0), 0, 50, 0, 100),
        freeThrowPct: normalizeValue(parsePercentage(stats.ftPct || stats.freeThrowPercentage || 0), 0, 95, 0, 100),

        // Advanced metrics (if available)
        per: stats.per ? normalizeValue(stats.per, 0, 30, 0, 100) : null,
        trueShootingPct: stats.tsPct ? normalizeValue(parsePercentage(stats.tsPct), 0, 70, 0, 100) : null
    };
}

/**
 * Position-specific stat weights
 */
function getPositionWeights(position) {
    const weights = {
        'PG': { // Point Guard - emphasis on assists and ball handling
            points: 0.20,
            assists: 0.30,
            rebounds: 0.10,
            steals: 0.15,
            blocks: 0.05,
            fieldGoalPct: 0.10,
            threePointPct: 0.05,
            freeThrowPct: 0.05
        },
        'SG': { // Shooting Guard - emphasis on scoring and perimeter shooting
            points: 0.30,
            assists: 0.15,
            rebounds: 0.10,
            steals: 0.15,
            blocks: 0.05,
            fieldGoalPct: 0.10,
            threePointPct: 0.10,
            freeThrowPct: 0.05
        },
        'SF': { // Small Forward - balanced all-around
            points: 0.25,
            assists: 0.15,
            rebounds: 0.15,
            steals: 0.15,
            blocks: 0.10,
            fieldGoalPct: 0.10,
            threePointPct: 0.05,
            freeThrowPct: 0.05
        },
        'PF': { // Power Forward - emphasis on rebounds and interior scoring
            points: 0.25,
            assists: 0.10,
            rebounds: 0.25,
            steals: 0.05,
            blocks: 0.15,
            fieldGoalPct: 0.15,
            threePointPct: 0.00,
            freeThrowPct: 0.05
        },
        'C': { // Center - emphasis on rebounds and blocks
            points: 0.20,
            assists: 0.05,
            rebounds: 0.30,
            steals: 0.05,
            blocks: 0.20,
            fieldGoalPct: 0.15,
            threePointPct: 0.00,
            freeThrowPct: 0.05
        }
    };

    // Default to SF weights if position unknown
    return weights[position] || weights['SF'];
}

/**
 * Apply position-specific adjustments
 */
function applyPositionAdjustments(rating, stats, position) {
    let adjusted = rating;

    // Bonus for elite efficiency
    const fgPct = parsePercentage(stats.fgPct || stats.fieldGoalPercentage || 0);
    if (fgPct > 55) {
        adjusted += 2; // Elite shooter bonus
    }

    // Bonus for high volume scorers
    const ppg = stats.ppg || stats.pointsPerGame || 0;
    if (ppg > 28) {
        adjusted += 3; // Superstar scorer bonus
    } else if (ppg > 25) {
        adjusted += 2;
    }

    // Bonus for triple-double threats
    const apg = stats.apg || stats.assistsPerGame || 0;
    const rpg = stats.rpg || stats.reboundsPerGame || 0;
    if (ppg > 15 && apg > 7 && rpg > 7) {
        adjusted += 4; // All-around excellence bonus
    }

    // Penalty for low efficiency
    if (fgPct < 40 && ppg > 15) {
        adjusted -= 3; // High volume, low efficiency penalty
    }

    return adjusted;
}

/**
 * Determine card rarity based on overall rating
 */
function getCardRarity(overallRating) {
    if (overallRating >= 90) return 'legendary';
    if (overallRating >= 85) return 'gold_rare';
    if (overallRating >= 80) return 'gold';
    if (overallRating >= 75) return 'silver_rare';
    if (overallRating >= 65) return 'silver';
    return 'bronze';
}

/**
 * Get card color based on rarity
 */
function getCardColor(rarity) {
    const colors = {
        'bronze': '#CD7F32',
        'silver': '#C0C0C0',
        'silver_rare': '#E8E8E8',
        'gold': '#FFD700',
        'gold_rare': '#FFA500',
        'legendary': '#9C27B0' // Purple for legendary
    };

    return colors[rarity] || colors.bronze;
}

/**
 * Get gradient colors for card background
 */
function getCardGradient(rarity) {
    const gradients = {
        'bronze': ['#8B4513', '#CD7F32', '#A0522D'],
        'silver': ['#808080', '#C0C0C0', '#A9A9A9'],
        'silver_rare': ['#C0C0C0', '#E8E8E8', '#D3D3D3'],
        'gold': ['#DAA520', '#FFD700', '#FFA500'],
        'gold_rare': ['#FF8C00', '#FFD700', '#FF6347'],
        'legendary': ['#4A148C', '#9C27B0', '#E91E63'] // Purple to pink
    };

    return gradients[rarity] || gradients.bronze;
}

/**
 * Calculate individual attribute ratings (for detailed cards)
 */
function calculateAttributeRatings(stats, position) {
    const normalized = normalizeStats(stats);

    return {
        // Offensive attributes
        scoring: Math.round((normalized.points * 0.6 + normalized.fieldGoalPct * 0.4)),
        shooting: Math.round((normalized.threePointPct * 0.6 + normalized.freeThrowPct * 0.4)),
        playmaking: Math.round(normalized.assists),

        // Physical attributes
        physicality: Math.round((normalized.rebounds * 0.6 + normalized.blocks * 0.4)),

        // Defensive attributes
        defense: Math.round((normalized.steals * 0.5 + normalized.blocks * 0.5)),

        // Overall efficiency
        efficiency: normalized.fieldGoalPct
    };
}

/**
 * Helper: Normalize value to target range
 */
function normalizeValue(value, sourceMin, sourceMax, targetMin, targetMax) {
    // Clamp source value
    const clamped = Math.max(sourceMin, Math.min(sourceMax, value));

    // Normalize to 0-1
    const normalized = (clamped - sourceMin) / (sourceMax - sourceMin);

    // Scale to target range
    return targetMin + normalized * (targetMax - targetMin);
}

/**
 * Helper: Parse percentage string to number
 */
function parsePercentage(value) {
    if (typeof value === 'number') return value;
    if (typeof value === 'string') {
        // Remove % sign if present
        const cleaned = value.replace('%', '').trim();
        const parsed = parseFloat(cleaned);
        return isNaN(parsed) ? 0 : parsed;
    }
    return 0;
}

// Import advanced stats calculator
const advancedStatsCalculator = require('./advancedStatsCalculator');

module.exports = {
    calculateOverallRating,
    getCardRarity,
    getCardColor,
    getCardGradient,
    calculateAttributeRatings,
    normalizeStats,
    // Export advanced stats functions
    calculateAdvancedAttributes: advancedStatsCalculator.calculateAdvancedAttributes,
    calculateAdvancedMetrics: advancedStatsCalculator.calculateAdvancedMetrics,
    getAttributeGrade: advancedStatsCalculator.getAttributeGrade,
    getAttributeColor: advancedStatsCalculator.getAttributeColor
};
